File: src/uniproxy/singbox/route.py
from __future__ import annotations

from typing import Literal, Sequence

from attrs import define, field

from uniproxy.uniproxy.base import BaseRule as UniproxyBaseRule
from uniproxy.uniproxy.rules import FinalRule as UniproxyFinalRule
from uniproxy.utils import maybe_map_to_tag, maybe_to_str

from .base import AbstractSingBox, BaseDnsServer, BaseInbound, BaseOutbound, BaseRuleSet
from .http_clients import HttpClient
from .route_rules import Rule


@define(hash=True)
class InlineRuleSet(BaseRuleSet):
    rules: Sequence[Rule]

    type: Literal["inline"] = "inline"


@define(hash=True)
class LocalRuleSet(BaseRuleSet):
    path: str

    format: Literal["binary", "source"]

    type: Literal["local"] = "local"

    def __attrs_post_init__(self) -> None:
        if self.format == "source" and not self.path.endswith("json"):
            raise ValueError(
                f"Invalid path for rule-set: {self.path}. Must end with '.json' for source format."
            )
        if self.format == "binary" and not self.path.endswith("srs"):
            raise ValueError(
                f"Invalid path for rule-set: {self.path}. Must end with '.srs' for binary format."
            )


@define(hash=True)
class RemoteRuleSet(BaseRuleSet):
    url: str
    format: Literal["binary", "source"] | None = None

    update_interval: float | None = None
    http_client: HttpClient | None = None

    # download_detour: BaseOutbound | str | None = field(
    #     default=None, converter=maybe_to_str
    # )
    # """
    # Tag of the outbound to download rule-set.

    # > Deprecated in sing-box 1.14.0
    # >
    # > `download_detour` is deprecated in sing-box 1.14.0 and will be removed in
    # > sing-box 1.16.0, use `http_client` instead.
    # """

    type: Literal["remote"] = "remote"

    def __attrs_post_init__(self) -> None:
        if self.format is None and self.url.endswith("json"):
            self.format = "source"
        elif self.format == "source" and not self.url.endswith("json"):
            raise ValueError(
                f"Invalid url for rule-set: {self.url}. Must end with '.json' for source format."
            )

        if self.format is None and self.url.endswith("srs"):
            self.format = "binary"
        elif self.format == "binary" and not self.url.endswith("srs"):
            raise ValueError(
                f"Invalid url for rule-set: {self.url}. Must end with '.srs' for binary format."
            )


type RuleSet = InlineRuleSet | LocalRuleSet | RemoteRuleSet


@define
class Route(AbstractSingBox):
    rules: Sequence[Rule]
    """List of [[Rule]]"""

    rule_set: Sequence[BaseRuleSet] | None = field(
        default=None, converter=maybe_map_to_tag
    )
    """List of [[rule-set]]"""

    final: BaseOutbound | str | None = field(default=None, converter=maybe_map_to_tag)
    """Default outbound tag. the first outbound will be used if empty."""

    auto_detect_interface: bool | None = None
    """
    > [!WARN] Only supported on Linux, Windows and macOS.

    Bind outbound connections to the default NIC by default to prevent routing loops under tun.

    Takes no effect if `outbound.bind_interface` is set.
    """

    override_android_vpn: bool | None = None
    """
    > [!WARN] Only supported on Android.

    Accept Android VPN as upstream NIC when `auto_detect_interface` enabled.
    """

    default_interface: str | None = None
    """
    > [!WARN] Only supported on Linux, Windows and macOS.

    Bind outbound connections to the specified NIC by default to prevent routing loops under tun.

    Takes no effect if `auto_detect_interface` is set.
    """

    default_mark: int | None = None
    """
    > [!WARN] Only supported on Linux.

    Set routing mark by default.

    Takes no effect if `outbound.routing_mark` is set.
    """

    default_http_client: HttpClient | str | None = field(
        default=None, converter=maybe_map_to_tag
    )
    """
    > [!NEW] Since sing-box 1.14.0

    Tag of the default HTTP Client used by remote rule-sets.

    If empty and `http_clients` is defined, the first HTTP client is used.
    """

    default_domain_resolver: BaseDnsServer | str | None = field(
        default=None, converter=maybe_to_str
    )
    """
    > [!NEW] Since sing-box 1.12.0

    Tag of target DNS server.
    """

    network_strategy: Literal["default", "hybrid", "fallback"] | None = None
    default_network_type: Literal["wifi", "cellular", "ethernet", "other"] | None = None

File: uniproxy/src/singbox/route.scala
package uniproxy.singbox.route

import upickle.default.ReadWriter

import uniproxy.singbox.abc.{AbstractRuleSet, AbstractSingBox, InboundLike, OutboundLike}
import uniproxy.singbox.typing.{RuleSetFormat, RuleSetType, SingBoxNetwork, SniffProtocol}

enum RuleSet(tag: String, format: RuleSetFormat, `type`: RuleSetType) extends ReadWriter {
  case LocalRuleSet(tag: String, format: RuleSetFormat, path: String)
      extends RuleSet(tag, format, RuleSetType.local)

  case RemoteRuleSet(
    tag: String,
    format: RuleSetFormat,
    url: String,
    download_detour: Option[InboundLike] = None,
    update_interval: Option[Float] = None,
  ) extends RuleSet(tag, format, RuleSetType.remote)
}

case class Rule(
  outbound: OutboundLike,
  inbound: Option[Seq[OutboundLike] | Seq[String]] = None,
  ip_version: Option["4" | "6"] = None,
  auth_user: Option[String | Seq[String]] = None,
  protocol: Option[SniffProtocol] = None,
  network: Option[SingBoxNetwork] = None,
  domain: Option[String | Seq[String]] = None,
  domain_suffix: Option[String | Seq[String]] = None,
  domain_keyword: Option[String | Seq[String]] = None,
  domain_regex: Option[String | Seq[String]] = None,
  ip_cidr: Option[String | Seq[String]] = None,
  ip_is_private: Option[Boolean] = None,
  source_ip_cidr: Option[Seq[String]] = None,
  source_ip_is_private: Option[Boolean] = None,
  source_port: Option[Int | Seq[Int]] = None,
  source_port_range: Option[String | Seq[String]] = None,
  port: Option[Int | Seq[Int]] = None,
  port_range: Option[String | Seq[String]] = None,
  rule_set: Option[RuleSet | Seq[RuleSet]] = None,
  rule_set_ip_cidr_match_source: Option[Boolean] = None,
  invert: Option[Boolean] = None,
) extends ReadWriter

// def from_uniproxy(cls, rule: UniproxyBasicRule | UniproxyGroupRule) -> Rule:
//     if not isinstance(rule, (UniproxyBasicRule, UniproxyGroupRule)):
//         raise ValueError(f"Expected UniproxyBasicRule, got {type(rule)}")

//     rule match :
//         case (
//             DomainRule(matcher=matcher, policy=policy)
//             | DomainGroupRule(matcher=matcher, policy=policy)
//         ):
//             return cls(outbound=String(policy), domain=matcher)  # type: ignore[reportArgumentType, arg-type]
//         case (
//             DomainSuffixRule(matcher=matcher, policy=policy)
//             | DomainSuffixGroupRule(matcher=matcher, policy=policy)
//         ):
//             return cls(outbound=String(policy), domain_suffix=matcher)  # type: ignore[reportArgumentType, arg-type]
//         case (
//             DomainKeywordRule(matcher=matcher, policy=policy)
//             | DomainKeywordGroupRule(matcher=matcher, policy=policy)
//         ):
//             return cls(outbound=String(policy), domain_keyword=matcher)  # type: ignore[reportArgumentType, arg-type]
//         case (
//             IPCidrRule(matcher=matcher, policy=policy)
//             | IPCidrGroupRule(matcher=matcher, policy=policy)
//             | IPCidr6Rule(matcher=matcher, policy=policy)
//             | IPCidr6GroupRule(matcher=matcher, policy=policy)
//         ):
//             return cls(outbound=String(policy), ip_cidr=matcher)  # type: ignore[reportArgumentType, arg-type]
//         case GeoIPRule(matcher=matcher, policy=policy):
//             # TODO: add extra opts to give a prefix or suffix
//             return cls(outbound=String(policy), rule_set=f"rs-geoip-{matcher}".lower())
//         case (
//             RuleSetRule(matcher, policy)
//             | DomainSetRule(matcher=matcher, policy=policy)
//         ):
//             matcher = String(matcher)
//             if matcher.startswith("http") and "://" in matcher:
//                 raise ValueError(
//                     f"Direct URL ({matcher}) is not supported currently while transforming from uniproxy external rule to sing-box rule"
//                 )
//             return cls(outbound=String(policy), rule_set=matcher)
//         case _:
//             raise ValueError(f"Unsupported rule type yet: {type(rule)}")

case class Route(
  /** List of route [[Rule]] */
  val rules: Seq[Rule],
  /** List of [[RuleSet]] */
  val rule_set: Option[Seq[RuleSet]] = None,
  /** Default outbound tag. the first outbound will be used if empty. */
  val `final`: Option[OutboundLike] = None,
  /**
   * > [!WARN] Only supported on Linux, Windows and macOS.
   *
   * Bind outbound connections to the default NIC by default to prevent routing
   * loops under tun.
   *
   * Takes no effect if `outbound.bind_interface` is set.
   */
  val auto_detect_interface: Option[Boolean] = None,
  /**
   * > [!WARN] Only supported on Android.
   *
   * Accept Android VPN as upstream NIC when `auto_detect_interface` enabled.
   */
  val override_android_vpn: Option[Boolean] = None,
  /**
   * > [!WARN] Only supported on Linux, Windows and macOS.
   *
   * Bind outbound connections to the specified NIC by default to prevent
   * routing loops under tun.
   *
   * Takes no effect if `auto_detect_interface` is set.
   */
  val default_interface: Option[String] = None,
  /**
   * > [!WARN] Only supported on Linux.
   *
   * Set routing mark by default.
   *
   * Takes no effect if `outbound.routing_mark` is set.
   */
  val default_mark: Option[Int] = None,
) extends AbstractSingBox derives ReadWriter
